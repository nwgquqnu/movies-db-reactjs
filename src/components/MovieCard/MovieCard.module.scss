%font {
    font-family: inherit;
    font-style: normal;
    font-weight: 500;
    color: #FFFFFF;
}

.movie-article {
    position: relative;
}

.movie-article:hover {
    cursor: pointer;
}

.movie-article img {
    margin: 0 auto;
    display: block;
}

.movie-details-btn {
    display: none;
    position: absolute;
    top: 16px;
    right: 16px;
    background: #2A202D;
    box-shadow: 0px 0px 4px rgba(0, 0, 0, 0.196596);
    border: none;
    border-radius: 50%;
    font-size: 0;
    width: 36px;
    height: 36px;
    padding: 0;
    background-image: radial-gradient(circle 1px at 50% 25%, white 2px, transparent 3px),radial-gradient(circle 1px at 50% 50%, white 2px, transparent 3px),radial-gradient(circle 1px at 50% 75%, white 2px, transparent 3px);
    cursor: pointer;
}

.movie-article:hover .movie-details-btn {
    display: block;
}

.movie-info {
    margin-top: 20px;
    display: flex;
    flex-flow: row wrap;
    justify-content: space-between;
}

.movie-title {
    font-size: 18px;
    line-height: 22px;

    mix-blend-mode: normal;
    opacity: 0.7;
}


.movie-year {
    font-size: 14px;
    line-height: 24px;
    
    text-align: center;
    
    opacity: 0.7;
    border: 1px solid rgba(151,151,151, 127);
    box-sizing: border-box;
    border-radius: 4px;

    width: 66px;
    height: 26px;
}

.movie-genre {
    font-size: 14px;
    line-height: 17px;

    mix-blend-mode: normal;
    opacity: 0.5;
    flex-basis: 100%;
}

.movie-genre, .movie-year, .movie-title {
    @extend %font;
    margin: 0;
}