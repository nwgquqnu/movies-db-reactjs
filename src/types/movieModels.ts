export interface NewMovie {
    // Movie title
    // example: La La Land
    title: string;
    // Movie tagline
    // example: Here's to the fools who dream.
    tagline: string;
    // Movie average raiting
    // example: 7.9
    vote_average: number;
    
    // Total count of votes for the movie
    // example: 6782
    vote_count:	number;
    
    // Movie release date
    // example: 2016-12-29
    release_date: string;
    
    // Url to the poster image
    // example: https://image.tmdb.org/t/p/w500/ylXCdC106IKiarftHkcacasaAcb.jpg
    poster_path: string;
    
    // Short description of the movie
    // example: Mia, an aspiring actress, serves lattes to movie stars in between auditions and Sebastian, a jazz musician, scrapes by playing cocktail party gigs in dingy bars, but as success mounts they are faced with decisions that begin to fray the fragile fabric of their love affair, and the dreams they worked so hard to maintain in each other threaten to rip them apart.
    overview: string;
    
    // Movie production budget
    // minimum: 0
    // example: 30000000
    budget:	number;
    
    // Movie revenue
    // example: 445435700
    // minimum: 0
    revenue: bigint;
    
    // Movie duration time
    // example: 128
    // minimum: 0
    runtime: number;
    
    genres: string[];
}



export type NewMovieKeys = keyof NewMovie;

type SameKeyAndValue<T> = {
    [Key in keyof T as Extract<Key, string>]: `${Extract<Key, string>}`;
};

export type NewMovieFields = SameKeyAndValue<NewMovie>;

type PickByType<T, Value> = {
    [P in keyof T as T[P] extends Value | undefined ? P : never]: T[P]
  }

export const newMovieFields = Object.freeze<NewMovieFields>({
    title: "title",
    tagline: "tagline",
    vote_average: "vote_average",
    vote_count: "vote_count",
    release_date: "release_date",
    poster_path: "poster_path",
    overview: "overview",
    budget: "budget",
    revenue: "revenue",
    runtime: "runtime",
    genres: "genres",
});

export const numericMovieFields = Object.freeze<SameKeyAndValue<PickByType<NewMovie, number> & PickByType<NewMovie, bigint>>>({
    vote_average: "vote_average",
    vote_count: "vote_count",
    budget: "budget",
    revenue: "revenue",
    runtime: "runtime",
})

export interface Movie extends NewMovie {
    id: string;
}

export enum SortOrder {
    ByNameAsc = "BY_NAME_ASC",
    ByNameDesc = "BY_NAME_DESC",
    ByReleaseAsc = "BY_RELEASE_ASC",
    ByReleaseDesc = "BY_RELEASE_DESC",
}

export enum UpdateActivity {
    addActivity = "ADD_ACTIVITY",
    editActivity = "EDIT_ACTIVITY",
    deleteActivity = "DELETE_ACTIVITY",
}

export interface MovieDbState {
    readonly moviesList: ReadonlyArray<Movie>;
    readonly filteredMoviesList: ReadonlyArray<Movie>;
    readonly genres: ReadonlyArray<string>;
    readonly activeGenre?: string;
    readonly selectedMovie?: Readonly<Movie>;
    readonly sortOrder: SortOrder;
    readonly movieUnderUpdateActivity?: Readonly<Movie>;
    readonly currentUpdateActivity?: UpdateActivity;
}
