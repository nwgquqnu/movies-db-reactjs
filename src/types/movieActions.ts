import { FetchedMoviesPayload, Movie } from "./movieModels";

export enum ActionType {
    ShowAddMovie = "SHOW_ADD_MOVIE",
    ShowDeleteMovie = "SHOW_DELETE_MOVIE",
    ShowEditMovie = "SHOW_EDIT_MOVIE",
    HideMovieUpdate = "HIDE_MOVIE_UPDATE",
    FetchedMovies = "FETCHED_MOVIES",
    FetchedGenres = "FETCHED_GENRES",

}

interface NoPayloadAction {
    type: ActionType.ShowAddMovie | ActionType.HideMovieUpdate;
}

interface EditMovieAction {
    type: ActionType.ShowEditMovie | ActionType.ShowDeleteMovie;
    payload: Movie;
}

export interface FetchedMoviesAction {
    type: ActionType.FetchedMovies;
    payload: FetchedMoviesPayload;
}

export interface FetchedGenresAction {
    type: ActionType.FetchedGenres;
    payload: ReadonlyArray<string>;
}

export type MovieDbStateAction = (
    EditMovieAction
    | NoPayloadAction
    | FetchedMoviesAction
    | FetchedGenresAction
);