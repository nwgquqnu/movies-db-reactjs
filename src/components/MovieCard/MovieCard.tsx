import * as React from 'react';
import { URLSearchParamsInit, useSearchParams } from 'react-router-dom';
import useSearchParamMovieId from '../../hooks/useSearchParamMovieId';
import { ActionType, MovieDbStateAction } from '../../types/movieActions';
import { Movie } from '../../types/movieModels';
import DetailsButton from '../DetailsButton';
import * as css from './MovieCard.module.scss';

interface MovieCardProps {
    movie: Movie;
    dispatch: React.Dispatch<MovieDbStateAction>;
}

function getYear(date: string): string {
    return date.slice(0, 4);
}

export default ({ movie, dispatch }: MovieCardProps) => {
    const [_, setSearchParamMovieId] = useSearchParamMovieId();

    return (
        <article className={css.movieArticle} onClick={() => setSearchParamMovieId(movie.id)}>
            <DetailsButton containerClassName={css.movieDetailsBtn} movie={movie} dispatch={dispatch} />
            <img src={movie.poster_path} alt={movie.title} />
            <div className={css.movieInfo}>
                <h1 className={css.movieTitle}>{movie.title}</h1>
                <p className={css.movieYear}>{getYear(movie.release_date)}</p>
                <p className={css.movieGenre}>{movie.genres.join(", ")}</p>
            </div>
        </article>
    );
};