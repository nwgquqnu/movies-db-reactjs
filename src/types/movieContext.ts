import * as React from "react";
import { MovieDbStateAction } from "./movieActions";
import { MovieDbState, SortOrder } from "./movieModels";

interface MovieContextObject {
    state: Readonly<MovieDbState>;
    dispatch: React.Dispatch<Readonly<MovieDbStateAction>>;
}

const defaultContextValue: MovieContextObject = {
    state: {
        moviesList: [],
        filteredMoviesList: [],
        genres: [],
        sortOrder: SortOrder.ByNameAsc,
    },
    dispatch: () => null,
};
const MovieContext = React.createContext(defaultContextValue);
export default MovieContext;