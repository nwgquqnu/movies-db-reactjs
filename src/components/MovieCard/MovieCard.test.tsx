import React from "react";
import { render, screen } from "@testing-library/react";
import MovieCard from ".";
import { Movie } from "../../types/movieModels";
import { MemoryRouter, Route, useLocation, Location, Routes, useSearchParams } from "react-router-dom";

describe("MovieCard", () => {
    let movie: Movie;

    beforeEach(() => {
        movie = {
            id: 123,
            title: "title",
            tagline: "tagline",
            vote_average: 0,
            vote_count: 1,
            release_date: "release_date",
            poster_path: "poster_path",
            overview: "overview",
            budget: 2,
            revenue: 3,
            runtime: 4,
            genres: ["genres"],
        };

    });

    it("corresponds to a snapshot", () => {
        const { asFragment } = render(
            <MovieCard movie={movie} dispatch={jest.fn()} />, { wrapper: MemoryRouter }
        );
        expect(asFragment()).toMatchSnapshot();
    });
    it("sets query params on click", () => {
        let testLocation: string | null = null;

        const LocationRetriever = () => {
            const [searchParams] = useSearchParams();
            testLocation = searchParams.toString();
            return <></>;
        };
        render(
            <MemoryRouter>
                <MovieCard movie={movie} dispatch={jest.fn()} />
                <Routes>
                    <Route
                        path="*"
                        element={<LocationRetriever/>}
                    />
                </Routes>
            </MemoryRouter>
        );
        const article = screen.queryByRole("article");
        expect(article).toBeInTheDocument();
        article?.click();

        expect(testLocation).toEqual("movie=123");
    });
});