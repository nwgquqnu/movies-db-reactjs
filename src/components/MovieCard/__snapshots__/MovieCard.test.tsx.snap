// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieCard corresponds to a snapshot 1`] = `
<DocumentFragment>
  <article>
    <nav
      class=""
      role="navigation"
    >
      <ul>
        <li>
          <button
            aria-haspopup="true"
          >
            Details
          </button>
          <ul
            aria-label="submenu"
            class="dropdown"
          >
            <li>
              <button>
                Edit
              </button>
            </li>
            <li>
              <button>
                Delete
              </button>
            </li>
          </ul>
        </li>
      </ul>
    </nav>
    <img
      alt="title"
      src="poster_path"
    />
    <div>
      <h1>
        title
      </h1>
      <p>
        rele
      </p>
      <p>
        genres
      </p>
    </div>
  </article>
</DocumentFragment>
`;
